applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_max_stress_response_function.cpp

```diff
@@ -0,0 +1,247 @@
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "adjoint_max_stress_response_function.h"
+#include "utilities/compare_elements_and_conditions_utility.h"
+
+namespace Kratos
+{
+    // --------------------------------------------------------------------------
+    AdjointMaxStressResponseFunction::AdjointMaxStressResponseFunction(ModelPart& rAdjointModelPart, Parameters ResponseSettings)
+    : AdjointStructuralResponseFunction(rAdjointModelPart, ResponseSettings),
+      mrAdjointModelPart(rAdjointModelPart),
+      mCriticalPartName(ResponseSettings["critical_part_name"].GetString())
+    {
+        mTracedStressType = StressResponseDefinitions::ConvertStringToTracedStressType(ResponseSettings["stress_type"].GetString());
+        mStressTreatment = StressResponseDefinitions::ConvertStringToStressTreatment(ResponseSettings["stress_treatment"].GetString());
+
+        if(ResponseSettings.Has("echo_level"))
+            mEchoLevel = ResponseSettings["echo_level"].GetInt();
+
+        KRATOS_ERROR_IF(mStressTreatment != StressTreatment::Mean)
+            << "AdjointMaxStressResponseFunction::AdjointMaxStressResponseFunction: Specified stress treatment not supported: " << ResponseSettings["stress_type"].GetString() << std::endl;
+
+        // Set default traced element in adjoint model part (first in the list of agglomerated elements).
+        ModelPart& adjoint_agglomeration_part = mrAdjointModelPart.GetSubModelPart(mCriticalPartName);
+        auto first_element = *(adjoint_agglomeration_part.ElementsBegin());
+        mpTracedElementInAdjointPart = mrAdjointModelPart.pGetElement(first_element.Id());
+        mpTracedElementInAdjointPart->SetValue(TRACED_STRESS_TYPE, static_cast<int>(mTracedStressType));
+    }
+
+    // --------------------------------------------------------------------------
+    AdjointMaxStressResponseFunction::~AdjointMaxStressResponseFunction(){}
+
+    // --------------------------------------------------------------------------
+    double AdjointMaxStressResponseFunction::CalculateValue(ModelPart& rPrimalModelPart)
+    {
+        KRATOS_TRY;
+
+        ModelPart& primal_agglomeration_part = rPrimalModelPart.GetSubModelPart(mCriticalPartName);
+
+        // Find max mean value and corresponding element in primal model part
+        double max_mean_stress = 0.0;
+        IndexType elem_id_at_max = 0;
+
+        for(auto& elem : primal_agglomeration_part.Elements())
+        {
+            Vector element_stress;
+            StressCalculation::CalculateStressOnGP(elem, mTracedStressType, element_stress, rPrimalModelPart.GetProcessInfo());
+
+            const SizeType stress_vec_size = element_stress.size();
+            double mean_stress = 0.0;
+            for(IndexType i = 0; i < stress_vec_size; ++i)
+                mean_stress += element_stress[i];
+            mean_stress /= stress_vec_size;
+
+            if(mean_stress > max_mean_stress)
```

Comment:
Is this definition of max stress reasonably? Let's imagine the following stresses values of a domain: (-10, 1, 1.5, 2). Would the better choice not to optimize for the element with stress value -10 instead of the element with 2?

Reply:
But that would not be a max() type of response function but a max(abs()), right? This is just a different way of "combinin" / "agglomerating" the stresses, much like the KS-function would be another option. Introducing agglomeration techniques would be very cool. But it is definitely a task for another PR.

Reply:
For VonMises stresses it is of course not a problem, but i do see Martins point.